packages/video_player/video_player_android/android/src/test/java/io/flutter/plugins/videoplayer/VideoAssetTest.java

```diff
@@ -0,0 +1,109 @@
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import android.net.Uri;
+import androidx.media3.common.MediaItem;
+import androidx.media3.datasource.DefaultHttpDataSource;
+import androidx.test.core.app.ApplicationProvider;
+import java.util.HashMap;
+import java.util.Map;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+
+/**
+ * Unit tests for {@link VideoAsset}.
+ *
+ * <p>This test suite <em>narrowly verifies</em> that the {@link VideoAsset} factory methods, {@link
+ * VideoAsset#fromRemoteUrl(String, VideoAsset.StreamingFormat, Map)} and {@link
+ * VideoAsset#fromAssetUrl(String)} follow the contract they have documented.
+ *
+ * <p>In other tests of the player, a fake asset is likely to be used.
+ */
+@RunWith(RobolectricTestRunner.class)
+public final class VideoAssetTest {
+  @Test
+  public void localVideoRequiresAssetUrl() {
+    assertThrows(
+        IllegalArgumentException.class,
+        () -> VideoAsset.fromAssetUrl("https://not.local/video.mp4"));
+  }
+
+  @Test
+  public void localVideoCreatesMediaItem() {
+    VideoAsset asset = VideoAsset.fromAssetUrl("asset:///asset-key");
+    MediaItem mediaItem = asset.getMediaItem();
+
+    assert mediaItem.localConfiguration != null;
+    assertEquals(mediaItem.localConfiguration.uri, Uri.parse("asset:///asset-key"));
+  }
+
+  private static DefaultHttpDataSource.Factory mockHttpFactory() {
+    DefaultHttpDataSource.Factory httpFactory = mock(DefaultHttpDataSource.Factory.class);
+    when(httpFactory.setUserAgent(anyString())).thenReturn(httpFactory);
+    when(httpFactory.setAllowCrossProtocolRedirects(anyBoolean())).thenReturn(httpFactory);
+    when(httpFactory.setDefaultRequestProperties(anyMap())).thenReturn(httpFactory);
+    return httpFactory;
+  }
+
+  @Test
+  public void remoteVideoByDefaultSetsUserAgentAndCrossProtocolRedirects() {
+    RemoteVideoAsset asset =
+        new RemoteVideoAsset(
+            "https://flutter.dev/video.mp4", VideoAsset.StreamingFormat.Unknown, new HashMap<>());
+
+    DefaultHttpDataSource.Factory mockFactory = mockHttpFactory();
+    asset.getMediaSourceFactory(ApplicationProvider.getApplicationContext(), mockFactory);
+
+    verify(mockFactory).setUserAgent("ExoPlayer");
+    verify(mockFactory).setAllowCrossProtocolRedirects(true);
+    verify(mockFactory, never()).setDefaultRequestProperties(anyMap());
+  }
+
+  @Test
+  public void remoteVideoOverridesUserAgentIfProvided() {
+    Map<String, String> headers = new HashMap<>();
+    headers.put("User-Agent", "FantasticalVideoBot");
+
+    RemoteVideoAsset asset =
+        new RemoteVideoAsset(
+            "https://flutter.dev/video.mp4", VideoAsset.StreamingFormat.Unknown, headers);
+
+    DefaultHttpDataSource.Factory mockFactory = mockHttpFactory();
+    asset.getMediaSourceFactory(ApplicationProvider.getApplicationContext(), mockFactory);
```

Comment:
Do you think it'd be helpful to test the overridden behavior of `getMediaSourceFactory` for `RemoteVideoAsset`s?

Reply:
So in production, there is a single method, `getMediaSourceFactory(Context)`, which always uses a new `DefaultHttpDataSource.Factory` (the same behavior before this patch).

For testing, we added a way to _not_ call `new DefaultHttpDataSource.Factory`, as we'd otherwise we'd not be able to assert that the HTTP factory is configured as expected.

I suspect I could assert that the factory makes a _real_ HTTP connection by default? I will leave a comment for the next person, and otherwise at least assert the returned object is real.

Reply:
Ah I see makes sense!